### PR TITLE
nixos/xsession: check for env vars

### PR DESCRIPTION
**Pull request by Kmeakin**
Read $XCOMPOSECACHE and $XDG_DATA_HOME environment variables when writing ~/.compose-cache and ~/.local/share, otherwise use default values.

###### Motivation for this change
Old script did not check for $XCOMPOSECACHE and $XDG_DATA_HOME
environment variables.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by peterhoeg** [comment]
Thank you for contributing to Nix.

Since you are using the XDG variables for the data path, how about ```XDG_CACHE_HOME``` for the default location of the compose cache (if xcomposecache isn't set)?

**Comment by Infinisil** [comment]
Aren't these env vars usually set up after this point and potentially in the sessions themselves? I for example have a ~/.xsession file that ends up setting the correct `XDG_DATA_HOME`, which happens here: https://github.com/NixOS/nixpkgs/blob/9137f83880a209b9e2e308b1a27486026e8aac4f/nixos/modules/services/x11/display-managers/default.nix#L132-L135
So this script doesn't even get the chance to see the correct env vars.

**Comment by Kmeakin** [comment]
@peterhoeg ~/.compose-cache is the default location, so if it has not been set by the user, other applications (e.g [compose (5)](https://www.x.org/archive/X11R7.5/doc/man/man5/Compose.5.html)) will read from ~/.compose-cache. This patch would have no effect unless the user specifically exports XCOMPOSECACHE.
I also think we should in general try not to change things from the upstream default, as it can cause confusion and inconsistent behavior between distros.

**Comment by Kmeakin** [comment]
@Infinisil The exact time that the environment variables are set depends on where they are defined. You are correct that, if they are set in `~/.xsession`, then they will not be available at the time the script is run. This can be solved by either:

- move the sourcing `~/.xsession` at the beginning
- defining your environment variables elsewhere, such as /etc/profile (this is can be used in NixOS by setting `environment.sessionVariables`. However this doesn't allow you to have different per-user values (e.g. user A wants `XDG_DATA_HOME` to be `~/.local/share`, and user B wants `XDG_DATA_HOME` to be `~/.hidden/local/share`)

**Comment by Infinisil** [comment]
The problem is that `~/.xsession` isn't a file you source, but rather execute. It represents the full user session, and it only exits once the user logs out.

**Comment by Kmeakin** [comment]
It also occurred to me that we could expose `services.xserver.displayManager.session.script` as an option. This would allow people with niche setups to override the default script. Currently `services.xserver.displayManager.sessionCommands` only appends to the default script.

**Comment by Kmeakin** [comment]
@Infinisil Apologies, I am not familiar with "manual" Xorg setups (I use KDE plasma). From looking at the xsession script, it seems like these env vars could be set in `~/.xprofile` if we move the sourcing of `~/.xprofile` to the beginning.

**Comment by Infinisil** [comment]
That might work

**Comment by Kmeakin** [comment]
@Infinisil Bump? Is there some proper way to request attention? I'm new to git/github, so sorry if this is the wrong way to go about it

**Comment by Kmeakin** [comment]
@Infinisil 
ping?

**Comment by Infinisil** [comment]
Not sure, maybe try pinging some people that were involved with this part of the code in the past

**Comment by peterhoeg** [comment]
As a KDE user, how would you go about overriding the location of the compose cache? Otherwise this looks fine to me but please run the various tests (sddm, plasma5, xfce, gnome3) to ensure this isn't breaking anything.

**Comment by Kmeakin** [comment]
@peterhoeg Personally, I set XCOMPOSECACHE in /etc/profile, though I imagine you could do it on a per-user basis with ~/.xprofile or ~/.pam_environment

**Comment by mmahut** [comment]
What is the status of this pull request?

**Comment by SFrijters** [comment]
While trying to declutter my `$HOME` I ran into this issue. It would be nice to have this rebased and merged at some point.

File: nixos/modules/services/x11/display-managers/default.nix [hunk 1]
```diff
@@ -110,22 +115,19 @@ let
 
       # Speed up application start by 50-150ms according to
       # http://kdemonkey.blogspot.nl/2008/04/magic-trick.html
-      rm -rf "$HOME/.compose-cache"
-      mkdir "$HOME/.compose-cache"
+      compose_cache="''${$XCOMPOSECACHE:-$HOME/.compose-cache}"
```
Comment (SFrijters) [inline nixos/modules/services/x11/display-managers/default.nix:118]: Shouldn't the `$` before XCOMPOSECACHE be dropped?

File: nixos/modules/services/x11/display-managers/default.nix [hunk 2]
```diff
@@ -110,22 +115,19 @@ let
...
+      rm -rf "$compose_cache"
+      mkdir -p "$compose_cache"
```
Comment (SFrijters) [inline nixos/modules/services/x11/display-managers/default.nix:120]: Maybe explicitly unset `compose_cache` when we're done with it?

**Pull request closed without merging**